src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterface.cs

```diff
@@ -1711,16 +1613,148 @@ private string ReadLineFromConsole(bool endOnTab, string initialContent, bool ca
+            {
+                m &= ~ConsoleControl.ConsoleModes.MouseInput;
+                m |= desiredMode;
+                ConsoleControl.SetMode(handle, m);
+            }
+
+            // If more characters are typed than you asked, then the next call to ReadConsole will return the
+            // additional characters beyond those you requested.
+            //
+            // If input is terminated with a tab key, then the buffer returned will have a tab (ascii 0x9) at the
+            // position where the tab key was hit.  If the user has arrowed backward over existing input in the line
+            // buffer, the tab will overwrite whatever character was in that position. That character will be lost in
+            // the input buffer, but since we echo each character the user types, it's still in the active screen buffer
+            // and we can read the console output to get that character.
+            //
+            // If input is terminated with an enter key, then the buffer returned will have ascii 0x0D and 0x0A
+            // (Carriage Return and Line Feed) as the last two characters of the buffer.
+            //
+            // If input is terminated with a break key (Ctrl-C, Ctrl-Break, Close, etc.), then the buffer will be
+            // the empty string.
+
+            _rawui.ClearKeyCache();
+            uint keyState = 0;
+            string s = string.Empty;
+            Span<char> inputBuffer = stackalloc char[MaxInputLineLength + 1];
+            if (initialContent.Length > 0)
+            {
+                initialContent.AsSpan().CopyTo(inputBuffer);
+            }
+
+            do
+            {
+                s += ConsoleControl.ReadConsole(handle, initialContent.Length, inputBuffer, MaxInputLineLength, endOnTab, out keyState);
+                Dbg.Assert(s != null, "s should never be null");
+
+                if (s.Length == 0)
+                {
+                    result = ReadLineResult.endedOnBreak;
+                    s = null;
+
+                    if (calledFromPipeline)
+                    {
+                        // make sure that the pipeline that called us is stopped
+
+                        throw new PipelineStoppedException();
+                    }
+
+                    break;
+                }
+
+                if (s.EndsWith(Crlf, StringComparison.Ordinal))
+                {
+                    result = ReadLineResult.endedOnEnter;
+                    s = s.Remove(s.Length - Crlf.Length);
+                    break;
+                }
+
+                int i = s.IndexOf(Tab, StringComparison.Ordinal);
+
+                if (endOnTab && i != -1)
+                {
+                    // then the tab we found is the completion character.  bit 0x10 is set if the shift key was down
+                    // when the key was hit.
+
+                    if ((keyState & 0x10) == 0)
```

Comment:
Is a `HasFlag()` appropriate here? Should this be a `[Flags()]` enum to make working with it easier?

Reply:
It is not clear in which direction we will move on, if we will.